src/System.Net.Http/src/System/Net/Http/Managed/HttpConnectionPools.cs

```diff
@@ -31,14 +31,25 @@ public HttpConnectionPools(int maxConnectionsPerServer)
         {
             _maxConnectionsPerServer = maxConnectionsPerServer;
             _pools = new ConcurrentDictionary<HttpConnectionKey, HttpConnectionPool>();
-            _cleaningTimer = new Timer(s => ((HttpConnectionPools)s).RemoveStalePools(), this, CleanPoolTimeoutMilliseconds, CleanPoolTimeoutMilliseconds);
+            // Start out with the timer not running, since we have no pools.
+            _cleaningTimer = new Timer(s => ((HttpConnectionPools)s).RemoveStalePools(), this, Timeout.Infinite, Timeout.Infinite);
         }
 
         /// <summary>Gets a pool for the specified endpoint, adding one if none existed.</summary>
         /// <param name="key">The endpoint for the pool.</param>
         /// <returns>The retrieved pool.</returns>
-        public HttpConnectionPool GetOrAddPool(HttpConnectionKey key) =>
-            _pools.GetOrAdd(key, (k, s) => new HttpConnectionPool(s), _maxConnectionsPerServer);
+        public HttpConnectionPool GetOrAddPool(HttpConnectionKey key)
+        {
+            // If we currently don't have any pools, a new one will be added for this
+            // connection. We should then start reaping stale pools.
+            if (_pools.IsEmpty)
```

Comment:
@stephentoub OK that was going to be my next question if there was a way we could separate out the Get and Add portions. Right now we don't know if a pool was added due to the use of `GetOrAdd`. I suppose we can use a `TryGet` and if that fails we know we need to add the pool and we can then do the check to start the timer as well.